src/tlshd/ktls.c

```diff
@@ -393,10 +395,52 @@ int tlshd_gnutls_priority_init(void)
 	pstring = tlshd_string_concat(pstring, ":-CIPHER-ALL");
 	if (!pstring)
 		return -ENOMEM;
+	pstring_sha256 = strdup(pstring);
+	if (!pstring_sha256) {
+		free(pstring);
+		return -ENOMEM;
+	}
+	pstring_sha384 = strdup(pstring);
+	if (!pstring_sha384) {
+		free(pstring_sha256);
+		free(pstring);
+		return -ENOMEM;
+	}
+
 	for (i = 0; i < ret; ++i) {
+		bool skip_sha256 = false;
+		bool skip_sha384 = false;
+
 		pstring = tlshd_cipher_string_emit(pstring, ciphers[i]);
-		if (!pstring)
+		if (!pstring) {
+			free(pstring_sha256);
+			free(pstring_sha384);
 			return -ENOMEM;
+		}
+		if (ciphers[i] == GNUTLS_CIPHER_AES_256_GCM)
+			skip_sha256 = true;
+		if (ciphers[i] == GNUTLS_CIPHER_AES_128_GCM)
+			skip_sha384 = true;
+		if (ciphers[i] == GNUTLS_CIPHER_AES_128_CCM)
+			skip_sha384 = true;
+		if (ciphers[i] == GNUTLS_CIPHER_CHACHA20_POLY1305)
+			skip_sha256 = true;
```

Comment:
Should this be `skip_sha384 = true`? It seems like ChaCha20 & Poly1305 are always used with SHA256.